compiler/noirc_evaluator/src/ssa/opt/remove_unreachable_instructions.rs

```diff
@@ -31,20 +41,88 @@ impl Function {
         // after an always failing one was found.
         let mut current_block_instructions_are_unreachable = false;
 
+        let one = self.dfg.make_constant(1_u32.into(), NumericType::bool());
+        let mut side_effects_condition = one;
+
         self.simple_reachable_blocks_optimization(|context| {
             let block_id = context.block_id;
 
             if current_block_id != Some(block_id) {
                 current_block_id = Some(block_id);
                 current_block_instructions_are_unreachable = false;
+                side_effects_condition = one;
             }
 
             if current_block_instructions_are_unreachable {
                 context.remove_current_instruction();
                 return;
             }
 
-            if always_fails(context.dfg, context.instruction()) {
+            let instruction = context.instruction();
+            if let Instruction::EnableSideEffectsIf { condition } = instruction {
+                side_effects_condition = *condition;
+                return;
+            };
+
+            let always_fails = match instruction {
+                Instruction::Constrain(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant != rhs_constant
+                }
+                Instruction::ConstrainNotEqual(lhs, rhs, _) => {
+                    let Some(lhs_constant) = context.dfg.get_numeric_constant(*lhs) else {
+                        return;
+                    };
+                    let Some(rhs_constant) = context.dfg.get_numeric_constant(*rhs) else {
+                        return;
+                    };
+                    lhs_constant == rhs_constant
+                }
+                Instruction::Binary(binary @ Binary { lhs, operator, rhs }) => {
+                    let requires_acir_gen_predicate =
+                        binary.requires_acir_gen_predicate(context.dfg);
+                    if requires_acir_gen_predicate {
+                        let predicate = context.dfg.get_numeric_constant(side_effects_condition);
+                        match predicate {
+                            Some(predicate) => {
+                                if predicate.is_zero() {
+                                    return;
```

Comment:
Do you think this part could be supported with some comments?

Reply:
Done!